### DISCUSSION TITLE
Why does a simple equals comparison fail?

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#13737] [General]
Here is the statement:
`if phd_info.device_name == name:'

Here are the values from print statements:
```
phd_info.device_name:  0BP10022000000
name:  0BP10022000000
phd_info.device_name == name  False
phd_info.device_name.strip() == name.strip()  False
type(phd_info.device_name)  <class 'str'>
type(name)  <class 'str'>
```
Yet somehow it is False. What is going on?

**Comment by robert-hh** [comment]
Maybe differently encoded characters, which look identical.

**Reply by robert-hh** [reply]
You could compare bytes(name, "utf-8") with bytes(phd_info.device_name, "utf-8"). That should show potential encodings.

**Reply by brianreinhold** [reply]
I think you are right. I looked at the source code for the aioble library call to see what was being handed up by the scanResults and it was a utf-8 formatted string (or bytes) - I am really confused by the myriad of byte notations that exist in Python and the `chr()` method returns unicode. However, the print statements for all these things looks the same. I was decoding the advertisement bytes myself using `chr()` and trying to compare it with what the library gave and though the LOOKED identical in print statements they refused to compare as equal. Lengths, classes, types, all seemed to be the same!

It cost me many many hours!

**Reply by mendenm** [reply]
@brianreinhold A useful thing to know is that `repr(something)` should print a uniquely identifiable representation of it. `repr(bytes_thing)` prints differently than `repr(string_thing)`.  `print()` itself uses an object's internal `__str__()` method, which is specifically designed to present a human readable, plain text version, so it hides the flavor of the incoming object.

**Reply by brianreinhold** [reply]
@mendenm Thanks for the tip. That is going to be really helpful. Would have saved me hours of confusion!
